### ISSUE TITLE
PROPOSAL: Add interfaces to Polly

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#257]
**TL;DR** Propose adding interfaces to Polly, in two segregated dimensions.  

+ Execution-interfaces to define execution overloads common to all resilience strategies.  
+ Orthogonal policy-kind-interfaces to define properties common to a kind of policy (eg the circuit state properties of all circuit-breaker policies).

Interface prototypes already visible in this branch: https://github.com/reisenberger/Polly/commits/feature/interfacesplusregistry

## Background

Prior to Polly v5.0, the team was reluctant to introduce interfaces as the execute method overloads were known to be in rapid evolution.  These have now stabilized.  

Additionally, clear guidance can be given that custom policies should be implemented by extending the Policy base class (intend to wiki/blog example) and _not_ by fulfilling interfaces.  This _mitigates_ one usual risk of introducing interfaces in an OSS project: that later evolving the interfaces risks user breakages.

## Execution interfaces

Execution interfaces [`ISyncPolicy`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/ISyncPolicy.cs), [`IAsyncPolicy`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/IAsyncPolicy.cs), [`ISyncPolicy<TResult>`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/ISyncPolicyTResult.cs) and [`IAsyncPolicy<TResult>`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/IAsyncPolicyTResult.cs) would define the execution overloads available to policies targeting sync/async, and non-generic / generic calls respectively.

Recent blog posts outline why Polly has [both non-generic and generic policies](http://www.thepollyproject.org/2017/06/07/why-does-polly-offer-both-non-generic-and-generic-policies/) and [separate sync and async policies](http://www.thepollyproject.org/2017/06/09/polly-and-synchronous-versus-asynchronous-policies/).

## Policy-kind interfaces

Orthogonal to the execution interfaces would be interfaces specific to each kind of Policy.  These would define properties and methods common to that type of policy, whether the policy instance targeted sync or async executions.  

For example, [`ICircuitBreakerPolicy`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/CircuitBreaker/ICircuitBreakerPolicy.cs) defines 

+ `CircuitState CircuitState`
+ `Exception LastException`
+ `void Isolate()`
+ `void Reset()`

with `ICircuitBreakerPolicy<TResult> : ICircuitBreakerPolicy` adding:

+ `TResult LastHandledResult`.

The benefit is that collections of these types of policies can then be managed - say, for monitoring purposes - as outlined by @christopherbahr in #205 .

[`BulkHead`](https://github.com/App-vNext/Polly/wiki/Bulkhead) is the only other Polly policy which exposes state, so [`IBulkheadPolicy`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/Bulkhead/IBulkheadPolicy.cs) is the only other similar interface with content.  However, other policy types would have similar empty marker interfaces (eg [`IRetryPolicy`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/Retry/IRetryPolicy.cs)) for parity/completeness.

## Registry support

The use of interfaces also supports the forthcoming [`PolicyRegistry` feature](https://github.com/App-vNext/Polly/pull/231), with a proposed empty [`IsPolicy`](https://github.com/reisenberger/Polly/blob/feature/interfacesplusregistry/src/Polly.Shared/IsPolicy.cs) marker interface identifying policies of all types.
